File: src/main/java/com/soapdogg/lispInterpreter/singleton/InterpreterSingleton.java
package com.soapdogg.lispInterpreter.singleton;

import com.soapdogg.lispInterpreter.interpreter.Interpreter;
import com.soapdogg.lispInterpreter.interpreter.RootNodePartitioner;
import lombok.Getter;

@Getter
public enum InterpreterSingleton {
    INSTANCE;

    private final RootNodePartitioner rootNodePartitioner;
    private final Interpreter interpreter;

    InterpreterSingleton() {
        rootNodePartitioner = new RootNodePartitioner();
        interpreter = new Interpreter(
            TokenizerSingleton.INSTANCE.getTokenizer(),
            ParserSingleton.INSTANCE.getRootParser(),
            EvaluatorSingleton.INSTANCE.getProgramEvaluator(),
            rootNodePartitioner,
            GeneratorSingleton.INSTANCE.getUserDefinedFunctionGenerator(),
            PrinterSingleton.INSTANCE.getListNotationPrinter()
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/singleton/InterpreterSingleton.kt
package com.soapdogg.lispInterpreter.singleton

import com.soapdogg.lispInterpreter.interpreter.Interpreter
import com.soapdogg.lispInterpreter.interpreter.RootNodePartitioner

enum class InterpreterSingleton {
    INSTANCE;

    val rootNodePartitioner: RootNodePartitioner = RootNodePartitioner()
    val interpreter: Interpreter = Interpreter(
        TokenizerSingleton.INSTANCE.tokenizer,
        ParserSingleton.INSTANCE.rootParser,
        rootNodePartitioner,
        GeneratorSingleton.INSTANCE.userDefinedFunctionGenerator,
        AsserterSingleton.INSTANCE.expressionListLengthAsserter,
        EvaluatorSingleton.INSTANCE.programEvaluator,
        PrinterSingleton.INSTANCE.listNotationPrinter
    )

}